### PR TITLE
Setup CSS basics

### PR DESCRIPTION
**Pull request by mattberridge**
Adds CSS infrastructure to the framework following quite a bit of experimentation and research regarding CSS modules, variables and theming. This took some time and configuration so I am writing it all down.

## Setup Notes

- CSS modules are setup using the Webpack [`css-loader`](https://github.com/webpack/css-loader) with [modules setting](https://github.com/webpack/css-loader#css-modules). Classnames are locally scoped and mangled with the following pattern `[name]__[local]___[hash:base64:5]`. **Make sure CSS files are called `ComponentName.css`** in the component folders, not `styles.css` or anything generic as the file name is taken as a class prefix (useful for debugging). [Read all about them](https://github.com/gajus/react-css-modules).
- So CSS can be required on the server I used the [`css-modules-require-hook`](https://github.com/css-modules/css-modules-require-hook) lib with settings / initialisation in [`babel.server.js`](https://github.com/iris-dni/iris-frontend/blob/206eb2c6e0ff57e8f5cb660f864632344ab7cb66/babel.server.js)
- Because **we are not using Sass** we needed a method for handling variables. There seem to be a lot of different libs but the cleanest I found seems to be [`css-modules-values`](https://github.com/css-modules/postcss-modules-values) so this is the one in place. You can import `@values` from other files easily and use them in CSS. There are [some caveats](https://github.com/css-modules/postcss-modules-values/issues) but we will see how it develops. A basic example can be found in [`Counter.css`](https://github.com/iris-dni/iris-frontend/blob/206eb2c6e0ff57e8f5cb660f864632344ab7cb66/src/components/Counter/Counter.css)
- I added [`postcss`](https://github.com/postcss/postcss) and [`postcss-autoprefixer`](https://github.com/postcss/autoprefixer) for vendor prefixing. Note the `browserslist` file in the root of the project which sets the autoprefixer settings in production. `postcss` lets us easily add other build steps for prod later, like [`cssnano`](http://cssnano.co/).
- As we will be developing an AZ theme on-top of an open source project default, I added the [`postcss-theme`](https://github.com/andywer/postcss-theme) lib to make this easier. We can define a default theme for now and then simply change the path in the [Webpack config](https://github.com/iris-dni/iris-frontend/blob/206eb2c6e0ff57e8f5cb660f864632344ab7cb66/webpack.config.prod.js#L16) when a new theme is added. Theme styles can be imported or composed i.e `@value black, white from theme(colors);` or `composes: menu from theme(components/menu);` as a shortcut to the base theme path, so refs don't need to be updated in the stylesheets when changing themes.
- Global styles are taken care of (e.g. CSS reset) in the [`src/styles/global.css`](https://github.com/iris-dni/iris-frontend/blob/206eb2c6e0ff57e8f5cb660f864632344ab7cb66/src/styles/global.css) file. These are required in the [`App.js` entry point](https://github.com/iris-dni/iris-frontend/blob/8f910060b8e4e0d9d30980d48cd5527e8b04435b/src/containers/App.js) so they are added to the bundle, before and in isolation from CSS modules which _only_ adds styles referenced in components. 
- The `ExtractTextPlugin` is used in production to [extract all CSS and bundle](https://github.com/gajus/react-css-modules#production) to a `styles.css` file. In development, CSS is injected into the `<head>` dynamically to work with hot-reloading - so you won't see any styling on server render for a split second before the client-side JS kicks in. To be clear: this won't happen on production.

## Linting

Also adds [stylelint](https://github.com/stylelint/stylelint) for linting using pre-commit hook and [`lint-staged`](https://github.com/okonet/lint-staged) lib for easier passing of only staged files for linting.

For `stylelint` I just added the core config [`stylelint-config-standard`](https://github.com/stylelint/stylelint-config-standard) for now as there is no _standard_ CSS linting convention at Espi - we can customise as we go,.

## TODOs

- [x] Compiled CSS is vendor prefixed
- [x] CSS reset is setup
- [x] CSS is compiled using CSS modules
- [x] Core styles defined as variables
- [x] Theme considerations and setup (core vs custom)
- [x] CSS linting is setup

## Move to a new story:

- [ ] Font loading considerations and setup (core vs custom)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,16 +5,17 @@
...
+    "build": "webpack --verbose --colors --progress --display-error-details --config webpack.config.prod.js",
+    "server": "NODE_PATH=\"./src\" node ./babel.server",
+    "server:dev": "webpack-dev-server --config webpack.config.dev.js",
     "nodecheckversion": "node ./scripts/nodecheckversion.js",
     "preinstall": "npm run nodecheckversion",
     "postinstall": "webpack -p --config ./webpack.config.prod.js",
-    "watch:client": "node ./node_modules/webpack/bin/webpack.js --verbose --colors --display-error-details --config webpack.config.dev.js && node ./node_modules/webpack-dev-server/bin/webpack-dev-server.js --config webpack.config.dev.js",
-    "dev": "npm run clean && node ./node_modules/concurrently/src/main.js --kill-others \"npm run watch:client\" \"npm run start\"",
-    "lint": "semistandard --verbose | snazzy",
+    "dev": "npm run clean && npm run build && concurrently --kill-others \"npm run server:dev\" \"npm run server\"",
```
Comment (mattberridge) [inline package.json]: @pgoetze note the changed commands for running dev

**Comment by mattberridge** [comment]
Ping @edenspiekermann/js

**Comment by nburkley** [comment]
Just curious as to why we're not using SASS?

File: dev-server.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+  publicPath: 'http://localhost:8080/dist/',
```
Comment (nburkley) [inline dev-server.js]: should this not be using an `ENV` var?

File: src/server/render.js [hunk 3]
```diff
@@ -27,9 +28,10 @@ export default (request, reply) => {
       );
 
       const stateJSON = asciiJSON.stringify(initialState).replace(/<\//g, '<\\/');
+
```
Comment (nburkley) [inline src/server/render.js]: extra space?

File: src/server/markup.js [hunk 4]
```diff
@@ -1,16 +1,20 @@
...
-    <html lang='en-us'>
+    <html lang='${config.lang}'>
       <head>
-        <meta charset='utf-8'>
-        <title>Hapi Universal Redux</title>
-        <link rel='shortcut icon' href='/favicon.ico'>
+        <meta charset='${config.charset}'>
+        <title>${config.title}</title>
+        ${stylesheet}
       </head>
```
Comment (mattberridge) [inline src/server/markup.js]: @artibella mentioned we might use [`react-helmet`](https://github.com/nfl/react-helmet) here later on, but we'll see

File: src/server/markup.js [hunk 5]
```diff
@@ -1,16 +1,20 @@
-export default (reactString, initialData, webserver) => {
+export default (reactString, initialData, webserver, config) => {
+  const stylesheet = process.env.NODE_ENV === 'production'
+    ? `<link rel='stylesheet' href='${webserver}/dist/styles.css'>`
+    : '';
```
Comment (pgoetze) [inline src/server/markup.js]: Is is a good practise to just do the logic here and or should the compilation of the stylesheet link be extracted as a `stylesheet(webserver)` function?

**Comment by mattberridge** [comment]
@nburkley because we (I? 😄 ) wanted to give CSS modules a go [for their benefits](http://glenmaddern.com/articles/css-modules) and they don't really work with a Sass setup (that I could configure). And actually Sass is pretty much overkill if you just use variables (now handled with post-css) and `composes` (CSS) instead of mixins / extend (Sass).

**Comment by pgoetze** [comment]
ping @HugoGiraudel

**Comment by nburkley** [comment]
@mattberridge Cool, thanks. Looking forward to hearing how it goes.

File: src/server/markup.js (hunk 5 above)
Comment (mattberridge) [inline src/server/markup.js]: I think you're right, I had this mind-marked as suboptimal but forgot to revisit. But we might use `react-helmet` later perhaps, which I will tackle with the font-loading story.

File: babel.server.js [hunk 6]
```diff
@@ -1,5 +1,12 @@
 require('babel-register');
 
+var cssModulesHook = require('css-modules-require-hook');
+
+cssModulesHook({
+  generateScopedName: '[name]__[local]___[hash:base64:5]',
```
Comment (HugoGiraudel) [inline babel.server.js]: Ideally, you want this in development, but a shorthash in production for better minification.

File: src/components/Counter/Counter.css [hunk 7]
```diff
@@ -0,0 +1,14 @@
+@value breakpoint, fontstack from 'styles/variables.css';
+@value black from theme(colors);
```
Comment (HugoGiraudel) [inline src/components/Counter/Counter.css]: How does the theming work exactly?

File: src/components/Counter/Counter.css [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+.root {
+  color: black;
+  font-family: fontstack, Arial, sans-serif;
```
Comment (HugoGiraudel) [inline src/components/Counter/Counter.css]: Please quote any font name.

File: src/components/Counter/Counter.css [hunk 9]
```diff
@@ -0,0 +1,14 @@
+@value breakpoint, fontstack from 'styles/variables.css';
+@value black from theme(colors);
+
+.root {
+  color: black;
```
Comment (HugoGiraudel) [inline src/components/Counter/Counter.css]: Please don’t use color keywords.

File: src/components/Counter/Counter.css [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  font-family: fontstack, Arial, sans-serif;
+  font-size: 30px;
```
Comment (HugoGiraudel) [inline src/components/Counter/Counter.css]: Please use `em` for font sizes.

File: src/components/Counter/Counter.css [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+@media breakpoint {
```
Comment (HugoGiraudel) [inline src/components/Counter/Counter.css]: Please rename or alias your breakpoint.

File: src/components/Counter/Counter.css [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+  .root {
+    color: #f00;
```
Comment (HugoGiraudel) [inline src/components/Counter/Counter.css]: Please if possible use `rgb(..)` or `hsl(..)` unless it goes against project team decision

File: src/components/Counter/Counter.scss [hunk 13]
```diff
@@ -0,0 +1,7 @@
+$color: black;
```
Comment (HugoGiraudel) [inline src/components/Counter/Counter.scss]: Please don’t use color keywords.

File: src/components/Counter/Counter.scss [hunk 14]
```diff
@@ -0,0 +1,7 @@
...
+$font-size: 30px;
```
Comment (HugoGiraudel) [inline src/components/Counter/Counter.scss]: Please use `em` for font sizes.

File: src/components/Counter/Counter.css (hunk 9 above)
Comment (mattberridge) [inline src/components/Counter/Counter.css]: `black` is the variable name imported above. it's just a proof of concept but I guess we should go with something like `colorBlack` for clarity

File: src/components/Counter/index.js [hunk 15]
```diff
@@ -1,4 +1,5 @@
 import React from 'react';
+import styles from './Counter.css';
```
Comment (HugoGiraudel) [inline src/components/Counter/index.js]: I would probably not capitalise CSS file names.

File: src/components/Counter/index.js [hunk 16]
```diff
@@ -12,10 +13,10 @@ export default class Counter extends React.Component {
     const { plus, minus } = this.props;
 
     return (
-      <div>
+      <div className={styles.root}>
         <span>Redux counter {count}</span>
         <div>
-          <button onClick={plus}>+</button>
+          <button onClick={plus} className={styles.button}>+</button>
           <button onClick={minus}>-</button>
```
Comment (HugoGiraudel) [inline src/components/Counter/index.js]: Missing a class here?

File: src/containers/App.js [hunk 17]
```diff
@@ -1,3 +1,5 @@
+import 'styles/global.css';
```
Comment (HugoGiraudel) [inline src/containers/App.js]: Ah. Interesting approach.

File: src/server/markup.js (hunk 5 above)
Comment (HugoGiraudel) [inline src/server/markup.js]: It should be extracted if possible.

File: src/components/Counter/Counter.css (hunk 7 above)
Comment (mattberridge) [inline src/components/Counter/Counter.css]: `theme(colors)` is basically just short for `/styles/theme/default/colors.css`. but it means as you go on, you don't have to refine loads of includes when changing your theme (just change base path in webpack)

File: src/styles/global.css [hunk 18]
```diff
@@ -0,0 +1,4 @@
+/**
+ * Imported in containers/App.js for all global styles
+ */
+@import './reset.css';
```
Comment (HugoGiraudel) [inline src/styles/global.css]: Are you sure you wanna go the reset route?

File: src/styles/reset.css [hunk 19]
```diff
@@ -0,0 +1,190 @@
+/**
+ * 1. Applying a proper box-model to the root element
+ *    From: https://css-tricks.com/inheriting-box-sizing-probably-slightly-better-best-practice/
+ * 2. Prevent iOS and IE text size adjust after device orientation change,
+ *    without disabling user zoom.
+ *    From: https://github.com/necolas/normalize.css/blob/2bdda84272650aedfb45d8abe11a6d177933a803/normalize.css#L11-L12
+ * 3. Prevent any horizontal overflow
+ */
+html {
+  box-sizing: border-box; /* 1 */
```
Comment (HugoGiraudel) [inline src/styles/reset.css]: What is your project decision on CSS declaration ordering?

File: src/styles/global.css (hunk 18 above)
Comment (HugoGiraudel) [inline src/styles/global.css]: Also you don’t need a reset if you go barbarian style (see file below).

File: src/styles/variables.css [hunk 20]
```diff
@@ -0,0 +1,2 @@
+@value breakpoint: (min-width: 900px);
+@value fontstack: 'Helvetica Neue', Arial, sans-serif;
```
Comment (HugoGiraudel) [inline src/styles/variables.css]: Please quote any font name.

File: src/components/Counter/Counter.css (hunk 8 above)
Comment (mattberridge) [inline src/components/Counter/Counter.css]: I thought it was just ones with spaces?

Comment (mattberridge) [inline src/components/Counter/Counter.css]: Never knew, thought it was just font names with spaces that needed quoted?

File: src/components/Counter/Counter.css (hunk 10 above)
Comment (mattberridge) [inline src/components/Counter/Counter.css]: Yeah, noted. This was just a quick proof of concept

File: src/components/Counter/Counter.css (hunk 11 above)
Comment (mattberridge) [inline src/components/Counter/Counter.css]: Yeah, noted. This was just a quick proof of concept

File: src/components/Counter/Counter.css (hunk 12 above)
Comment (mattberridge) [inline src/components/Counter/Counter.css]: Yeah, noted. This was just a quick proof of concept

File: src/components/Counter/Counter.scss (hunk 14 above)
Comment (mattberridge) [inline src/components/Counter/Counter.scss]: Yeah, noted. This was just a quick proof of concept

File: src/components/Counter/Counter.css (hunk 7 above)
Comment (HugoGiraudel) [inline src/components/Counter/Counter.css]: Interesting. Looking forward to seeing how it works.

File: src/components/Counter/Counter.css (hunk 9 above)
Comment (HugoGiraudel) [inline src/components/Counter/Counter.css]: Definitely.

File: src/components/Counter/Counter.css (hunk 8 above)
Comment (HugoGiraudel) [inline src/components/Counter/Counter.css]: Even font names with spaces do not need to be quoted. :)
So either you quote them all, or you quote none of them.

File: babel.server.js (hunk 6 above)
Comment (mattberridge) [inline babel.server.js]: Example?

File: dev-server.js (hunk 2 above)
Comment (mattberridge) [inline dev-server.js]: @pgoetze thoughts?

Comment (pgoetze) [inline dev-server.js]: I think it makes sense to provide a decent default for the dev server, but
we could at least make the `http://localhost:8080` overwritable by an ENV variable I think. Sth like this?

```
`${process.env.DEV_HOST || 'http://localhost:8080'}/dist/`
```

(or whatever might be a better name for `DEV_HOST`)

Comment (mattberridge) [inline dev-server.js]: This touches multiple places and raises a bigger question of installation / config. Let's tackle it later on.

**Pull request merged**